specs/fizzbee/LeaderFollower/VeniceLeaderFollower.fizz

```diff
@@ -0,0 +1,136 @@
+
+# Init is a special action called when the model starts. It is used to initialize the state of the model.
+action Init:
+    # Currently, FizzBee does not support parameters. So, we have to define these as variables with uppercase.
+    KEYS = ('key1', 'key2')
+    VALUES = ('value1', 'value2')
+    LEADER = 'LEADER'
+    FOLLOWER = 'FOLLOWER'
+    N_NODES = 3
+    NODE_IDS = range(N_NODES)
+
+    # Initialize the state of each node
+    real_time_topic = []
+    version_topic = []
+    nodes = [
+        {
+            'state': 'FOLLOWER',
+            'rtOffset': 0,
+            'vtOffset': 0,
+            'persistedRecords': {}
+        }
+        for i in NODE_IDS
+    ]
+
+
+# The action keyword is used to define an action. The atomic keyword implies
+# there are no yield points within the statements.
+atomic action PromoteLeader:
+    # A FOLLOWER ca be promoted to be a LEADER if and only if there is no leader in the system.
+
+    if [node for node in nodes if node['state'] == 'LEADER']:
+        return
+    # `any` is a special keyword that creates non-determinism or alternate timelines.
+    # This is similar in structure to `for`, but in each timeline, only one of the
+    # non-deterministic choices is executed.
+    any id in NODE_IDS:
+        nodes[id]['state'] = 'LEADER'
+
+
+atomic action DemoteLeader:
+    any id in NODE_IDS:
+        if nodes[id]['state'] == 'LEADER':
+            nodes[id]['state'] = 'FOLLOWER'
+
+
+atomic action ClientProducesToVenice:
+    any kv in [(k, v) for k in KEYS for v in VALUES]:
+        real_time_topic.append(kv)
+
+
+# The fair keyword implies this action is weakly fair. That is, the action is always enabled,
+# it will eventually happen.
+atomic fair action LeaderConsume:
+  any id in NODE_IDS:
+    if nodes[id]['state'] == 'LEADER':
+        if nodes[id]['vtOffset'] >= len(version_topic):
+            real_time_consume(id)
+        else:
+            version_topic_consume(id)
+
+
+# The strong modifier to the fair keyword implies this action is strongly fair. That is, the action
+# gets enabled repeatedly (that is it may get disabled in between), it will eventually happen.
+atomic fair<strong> action FollowerConsume:
+    any id in NODE_IDS:
+        if nodes[id]['state'] == 'FOLLOWER' and nodes[id]['vtOffset'] < len(version_topic):
+            version_topic_consume(id)
+
+
+atomic func real_time_consume(nodeId):
+    # A leader consuming unprocessed msgs from the real-time topic.
+
+    # Check if the node's offset is within the bounds of the real-time topic
+    if nodes[nodeId]['rtOffset'] >= len(real_time_topic):
+        return
+
+    # Update the node's state and persisted records
+    offset = nodes[nodeId]['rtOffset']
+    key, value = real_time_topic[offset][0], real_time_topic[offset][1]
+
+    nodes[nodeId]['rtOffset'] += 1
+    nodes[nodeId]['persistedRecords'][key] = value
+
+    # Append the consumed event to the version topic
+    version_topic.append((key, value, offset))
```

Comment:
This may be is a minor detail, but in today's implementation, the order of operation is: VT append, local persistence, update RT offset. And the reason why we do it this way is so that we have idempotency in case we crash at any point in between the operations (and so, perhaps this function should _not_ have the `atomic` keyword?)

We are actually considering changing the ordering of these operations (and it might be interesting to model it both ways), but that change (#910) is not integrated (or even necessarily fully agreed upon...) yet.

Reply:
So actually, you touch on a really interesting facet that would probably tell us that we should consider either a refinement of the leader follower spec, or, tweaking it a bit.  The way TLA+ works is that you define states.  Currently, the model is very simplistic.  It defines the leader actions of consuming from RT, updating local state, and producing to VT as a single state.

Reply:
I can change this, but to model it correctly, I need one additional information. How do you handle duplicate message publication to the topic?
Kafka doesn't have a built deduplication like that of SQS or other queueing services. But I know Kafka streams and client side libraries handle it in a way this never becomes an issue. 
Let me know, how should I model that? Can I assume the deduplication is somehow handled without actually implementing it?

The scenario:
1. RealTimeConsume
    i. VT append ((k1,v1,0))
    ii. Crash
2. RealTimeConsume
    i. VT append (k1,v1,0)
    ii. Crash
This can go on infinitely (leading to infinite state space)
So, I could quickly add a way to dedupe if an exact message is sent, the append would ignore and we will proceed with the remaining steps.

Correct me if I missed something.

Reply:
So in Venice the way we essentially deal with it is we have write collision resolution (last writer wins).  So I think your suggestion of append ignore is the right way to go.